📖 Document multi-tenancy contract (#4074)
kubernetes-sigs/cluster-api: docs/book/src/developer/architecture/controllers/multi-tenancy.md

# Multi tenancy

Multi tenancy in Cluster API defines the capability of an infrastructure provider to manage different credentials, each
one of them corresponding to an infrastructure tenant.

## Contract
In order to support multi tenancy, the following rule applies:

> - All the infrastructure provider are required to manage different credentials.

Reviewer: ```suggestion
- Infrastructure providers MUST be able to manage different sets of credentials (if any).
```

> - All the provider should deploy and run web-hook and with managers;
>   please note this is the default in controllers generated with kubebuilder. 

Reviewer: ```suggestion
- Providers SHOULD deploy and run any kind of webhook (validation, admission, conversion) following Cluster API codebase best practices for the same release. 
```

> - All the provider are required to create a `component.yaml` accordingly.

Author: My intent is to link documentation from https://github.com/kubernetes-sigs/cluster-api/pull/3985 here, depending on which PR is going to merge first

> - All the provider are required to create a `component.yaml` accordingly.

Reviewer: ```suggestion
- Providers MUST create and publish a `{type}-components.yaml` accordingly.
```
This should be the case today? We get infrastructure-components, bootstrap, etc?

Author: yes!
